application/F3DStarter.cxx

```diff
@@ -165,6 +166,92 @@ class F3DStarter::F3DInternals
     }
   }
 
+  void addOutputImageMetadata(f3d::image& image)
+  {
+    std::stringstream cameraMetadata;
+    {
+      const auto state = Engine->getWindow().getCamera().getState();
+      const auto vec3toJson = [](const std::array<double, 3>& v)
+      {
+        std::stringstream ss;
+        ss << "[" << v[0] << ", " << v[1] << ", " << v[2] << "]";
+        return ss.str();
+      };
+      cameraMetadata << "{\n";
+      cameraMetadata << "  \"pos\": " << vec3toJson(state.pos) << ",\n";
+      cameraMetadata << "  \"foc\": " << vec3toJson(state.foc) << ",\n";
+      cameraMetadata << "  \"up\": " << vec3toJson(state.up) << ",\n";
+      cameraMetadata << "  \"angle\": " << state.angle << "\n";
+      cameraMetadata << "}\n";
+    }
+
+    image.setMetadata("camera", cameraMetadata.str());
+  }
+
+  /**
+   * Substitute the following variables in a filename template:
+   * - `{app}`: application name (eg. `F3D`)
+   * - `{version}`: application version (eg. `2.4.0`)
+   * - `{version_full}`: full application version (eg. `2.4.0-abcdefgh`)
+   * - `{model}`: current model filename without extension (eg. `foo` for `/home/user/foo.glb`)
+   * - `{model.ext}`: current model filename with extension (eg. `foo.glb` for `/home/user/foo.glb`)
+   * - `{n}`: auto-incremented number to make filename unique
+   */
+  std::filesystem::path applyFilenameTemplate(const std::string& filenameTemplate)
+  {
+    std::string filename = filenameTemplate;
+
+    {
+      std::regex re("\\{app\\}");
+      filename = std::regex_replace(filename, re, F3D::AppName);
+    }
+    {
+      std::regex re("\\{version\\}");
+      filename = std::regex_replace(filename, re, F3D::AppVersion);
+    }
+    {
+      std::regex re("\\{version_full\\}");
+      filename = std::regex_replace(filename, re, F3D::AppVersionFull);
+    }
+
+    {
+      std::regex re("\\{model\\}");
+      if (std::regex_search(filename, re))
+      {
+        const auto file = FilesList[CurrentFileIndex];
+        filename = std::regex_replace(filename, re, file.stem().string());
+      }
+    }
+    {
+      std::regex re("\\{model[.]ext\\}");
+      if (std::regex_search(filename, re))
+      {
+        const auto file = FilesList[CurrentFileIndex];
+        filename = std::regex_replace(filename, re, file.filename().string());
+      }
+    }
+
+    {
+      /* try and find non-existing filename; must be done last, assumes full path */
+      std::regex re("\\{n\\}");
+      if (std::regex_search(filename, re))
+      {
+        std::string fn = std::regex_replace(filename, re, "1");
+        for (size_t i = 2; std::filesystem::exists(fn) && i < 1000000; ++i)
```

Comment:
`1e6` instead ?

Reply:
`1e6` is a `double` tho, does that matter?
Alternatively the internet tells me C++14 can do `1'000'000`

Reply:
nice, sounds good to me.

Reply:
`1'000'000` won't fly with `clang-format` and `Standard: Cpp03`... need to upgrade the config

Reply:
alright, lets keep 1000000